File: Mika/tags/PKG-adhoco-4/core-vm/java/java/io/BufferedReader.java
/**************************************************************************
* Copyright (c) 2001 by Punch Telematix. All rights reserved.             *
*                                                                         *
* Redistribution and use in source and binary forms, with or without      *
* modification, are permitted provided that the following conditions      *
* are met:                                                                *
* 1. Redistributions of source code must retain the above copyright       *
*    notice, this list of conditions and the following disclaimer.        *
* 2. Redistributions in binary form must reproduce the above copyright    *
*    notice, this list of conditions and the following disclaimer in the  *
*    documentation and/or other materials provided with the distribution. *
* 3. Neither the name of Punch Telematix nor the names of                 *
*    other contributors may be used to endorse or promote products        *
*    derived from this software without specific prior written permission.*
*                                                                         *
* THIS SOFTWARE IS PROVIDED ``AS IS'' AND ANY EXPRESS OR IMPLIED          *
* WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF    *
* MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.    *
* IN NO EVENT SHALL PUNCH TELEMATIX OR OTHER CONTRIBUTORS BE LIABLE       *
* FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR            *
* CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF    *
* SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR         *
* BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,   *
* WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE    *
* OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN  *
* IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.                           *
**************************************************************************/

/*
** $Id: BufferedReader.java,v 1.2 2006/10/04 14:24:14 cvsroot Exp $
*/

package java.io;

public class BufferedReader extends Reader {

  private Reader in;

  /*
  ** these 5 fields have the same meaning as in the BufferedInputStream...
  */

  private char buf[];
  private int pos;
  private int count;
  private int markpos;
  private int marklimit;

  public BufferedReader(Reader in) {
    this(in, 8192);
  }
  public BufferedReader(Reader in, int size) {
    super(in);
    if (size <= 0){
     throw new IllegalArgumentException("negative buffer size");
    }
    this.in = in;
    buf = new char[size];
    markpos = -1;
  }

  public int read() throws IOException {
    synchronized (lock) {
      if(buf == null){
        throw new IOException("Reader is closed");
      }
      if(count <= pos && !updateBuffer()){
        return -1;
      }
      return buf[pos++];
    }
  }

  public int read(char b[], int off, int len) throws IOException {
    synchronized (lock) {
      if(buf == null){
        throw new IOException("Reader is closed");
      }
      if(off < 0 || len < 0 || off > b.length - len){
        throw new IndexOutOfBoundsException();
      }
      int rd = 0;
      while(len > 0) {
        //first take the bytes out of buf
        if(count > pos){
          int cp = count - pos;
          if(len > cp){
            System.arraycopy(buf, pos, b, off, cp);
            off += cp;
            len -= cp;
            pos += cp;
            rd +=cp;
          }else{
            System.arraycopy(buf, pos, b, off, len);
            pos += len;
            break;
          }
        }
        //we need more bytes ...
        if(len > buf.length && (markpos == -1 || pos > (markpos+marklimit))){
          len = in.read(b, off, len);
          if(len == -1 && rd > 0){
            len = 0;
          }
          break;
        }else if (!updateBuffer()){
          len = (rd == 0 ? -1 : 0);
          break;
        }
      }
      return rd + len;
    }
  }

  private native int locateEnd(char[] chars, int startPos, int endPos);

  public String readLine() throws IOException {
    synchronized (lock) {
      if(buf == null){
        throw new IOException("Reader is closed");
      }
      StringBuffer line = new StringBuffer(64);
      int end;
      do {
        end = locateEnd(buf,pos,count);
        if(end != -1){
          line.append(buf,pos, end - pos);
          pos = end;
          if(buf[pos++] == '\r'){
            if ((count > pos || updateBuffer()) &&  buf[pos] == '\n'){
              pos++;
            }
          }
          return line.toString();
        }
        else {
          line.append(buf,pos, count - pos);
          pos = count;
        }
      } while(updateBuffer());
      if (line.length() == 0 && end == -1){
        return null;
      }
      return line.toString();
    }
  }

  public long skip(long n) throws IOException {
    synchronized (lock) {
      if(buf == null){
        throw new IOException("Reader is closed");
      }
      if(n < 0){
        throw new IllegalArgumentException();
      }
      long skipped = 0;
      while(n > 0){
        if((n - count + pos) > 0){
          //not enough bytes in the buffer
          int skip = count - pos;
          pos = count;
          if((n - skip) > buf.length && (markpos == -1 || pos > (markpos+marklimit))){
            return in.skip(n-skip)+skip;
          }else if(!updateBuffer()){
            break;
          }
        }else{
          //take bytes from the buffer
          pos += (int)n;
          skipped += n;
          break;
        }
      }
      return skipped;
    }
  }

  public boolean ready() throws IOException {
    synchronized (lock) {
      if(buf == null){
        throw new IOException("Reader is closed");
      }
      return (count > pos || in.ready());
    }
  }

  public boolean markSupported() {
    return true;
  }

  public void mark(int readLimit) throws IOException {
    synchronized (lock) {
      if(buf == null){
        throw new IOException("Reader is closed");
      }
      if(readLimit < 0){
        throw new IllegalArgumentException();
      }
      marklimit = readLimit;
      markpos = pos;
    }
  }

  public void reset() throws IOException {
    synchronized (lock) {
      if(buf == null || markpos == -1 || (markpos+marklimit < pos)){
        throw new IOException();
      }
      pos = markpos;
    }
  }

  public void close() throws IOException {
    synchronized (lock) {
      if(buf != null){
        buf = null;
        in.close();
        in = null;
      }
    }
  }

  private boolean updateBuffer() throws IOException{
    if(markpos == -1 || pos >= (markpos + marklimit)){
      int rd = in.read(buf, 0, buf.length);
      pos = 0;
      markpos = -1;
      if (rd == -1){
        count = 0;
        return false;
      }
      else if (rd < 1) {
        new Exception("OOPS - read on "+in+" returned "+rd).printStackTrace();
      }
      count = rd;
      return true;
    }else{
      //a position is marked and still valid
      char[] chars = buf;
      if (marklimit > buf.length){
         //let the buffer grow
         chars = new char[marklimit];
      }
      int cp = count - markpos;
      System.arraycopy(buf, markpos, chars, 0 ,cp);
      int rd = in.read(chars, cp, chars.length - cp);
      pos = cp;
      markpos = 0;
      buf = chars;
      if (rd == -1){
        count = cp;
        return false;
      }
      else if (rd < 1) {
        new Exception("OOPS - read on "+in+" returned "+rd).printStackTrace();
      }
      count = cp + rd;
      return true;
    }
  }
}

File: Mika/trunk/core-vm/src/native/java/io/BufferedReader.c
/**************************************************************************
* Parts copyright (c) 2001 by Punch Telematix.  All rights reserved.      *
* Parts copyright (c) 2010 by Chris Gray, /k/ Embedded Java Solutions.    *
* All rights reserved.                                                    *
*                                                                         *
* Redistribution and use in source and binary forms, with or without      *
* modification, are permitted provided that the following conditions      *
* are met:                                                                *
* 1. Redistributions of source code must retain the above copyright       *
*    notice, this list of conditions and the following disclaimer.        *
* 2. Redistributions in binary form must reproduce the above copyright    *
*    notice, this list of conditions and the following disclaimer in the  *
*    documentation and/or other materials provided with the distribution. *
* 3. Neither the name of Punch Telematix or of /k/ Embedded Java Solutions*
*    nor the names of other contributors may be used to endorse or promote*
*    products derived from this software without specific prior written   *
*    permission.                                                          *
*                                                                         *
* THIS SOFTWARE IS PROVIDED ``AS IS'' AND ANY EXPRESS OR IMPLIED          *
* WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF    *
* MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.    *
* IN NO EVENT SHALL PUNCH TELEMATIX, /K/ EMBEDDED JAVA SOLUTIONS OR OTHER *
* CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,   *
* EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,     *
* PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR      *
* PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF  *
* LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING    *
* NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS      *
* SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.            *
**************************************************************************/

#include <stdio.h>

#include "core-classes.h"
#include "arrays.h"
#include "exception.h"
#include "fields.h"
#include "interpreter.h"
#include "heap.h"
#include "jni.h"
#include "locks.h"
#include "methods.h"
#include "threads.h"

static jclass   class_BufferedReader;
static jmethodID updateBuffer_method;

w_int BufferedReader_locateEnd(JNIEnv *env, w_instance this, w_instance Array, w_int start, w_int stop) {

  if(Array){
    w_int length = instance2Array_length(Array);
    w_char* chars  = instance2Array_char(Array);
    if(length >= stop && start >= 0){
      while(start < stop){
        if(chars[start] == 10 || chars[start] == 13){
          woempa(1, "found end-of-line character %02x at position[%d]\n", chars[start], start);
          return start;
        }
        start++;
      }
    }
  }

  woempa(1, "found no end-of-line character\n");

  return -1;
}

w_int BufferedReader_read(JNIEnv *env, w_instance thisBufferedReader) {
  w_thread thread = JNIEnv2w_thread(env);
  w_int result = -1;
  x_monitor m;
  w_instance lock = getReferenceField(thisBufferedReader, F_Reader_lock);
  w_instance buf;
  w_frame new_frame;

  m = getMonitor(lock);
  x_monitor_eternal(m);
  buf = getReferenceField(thisBufferedReader, F_BufferedReader_buf);

  if (!buf) {
    x_monitor_exit(m);
    throwException(thread, clazzIOException, "BufferedReader is closed");

    return -1;
  }

  w_word *posptr = wordFieldPointer(thisBufferedReader, F_BufferedReader_pos);
  w_int count = getIntegerField(thisBufferedReader, F_BufferedReader_count);
  w_boolean updated;

  if (count <= (w_int)*posptr) {
    if (!updateBuffer_method) {
      class_BufferedReader = clazz2Class(clazzBufferedReader);
      updateBuffer_method = (*env)->GetMethodID(env, class_BufferedReader, "updateBuffer", "()Z"); 
      woempa(7,"updateBuffer_method is %M\n",updateBuffer_method);
    }
    new_frame = activateFrame(thread, updateBuffer_method, 0, 1, thisBufferedReader, stack_trace);
    updated = (w_boolean) new_frame->jstack_top[-1].c;
    deactivateFrame(new_frame, NULL);
    removeLocalReference(thread, thisBufferedReader);
    if (!updated) {
      x_monitor_exit(m);

      return -1;
    }
  }
  // The buf array may very well have changed as a result of updateBuffer!
  buf = getReferenceField(thisBufferedReader, F_BufferedReader_buf);
  result = instance2Array_char(buf)[*posptr];
  ++(*posptr);
  x_monitor_exit(m);

  return result;
}

